official/vision/beta/projects/yolo/optimization/sgd_torch.py

```diff
@@ -0,0 +1,281 @@
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import array_ops
+from tensorflow.keras.optimizers.schedules import LearningRateSchedule
+from tensorflow.python.training import gen_training_ops
+
+import tensorflow as tf
+import logging
+
+__all__ = ['SGDTorch']
+
+
+def _var_key(var):
+  """Key for representing a primary variable, for looking up slots.
+  In graph mode the name is derived from the var shared name.
+  In eager mode the name is derived from the var unique id.
+  If distribution strategy exists, get the primary variable first.
+  Args:
+    var: the variable.
+  Returns:
+    the unique name of the variable.
+  """
+
+  # pylint: disable=protected-access
+  # Get the distributed variable if it exists.
+  if hasattr(var, "_distributed_container"):
+    var = var._distributed_container()
+  if var._in_graph_mode:
+    return var._shared_name
+  return var._unique_id
+
+
+class SGDTorch(tf.keras.optimizers.Optimizer):
+  """Optimizer that simulates the SGD module used in pytorch. 
+  
+  
+  For details on the differences between the original SGD implemention and the 
+  one in pytorch: https://pytorch.org/docs/stable/generated/torch.optim.SGD.html.
+  This optimizer also allow for the usage of a momentum warmup along side a 
+  learning rate warm up, though using this is not required. 
+
+  Example of usage for training:
+  ```python
+  opt = SGDTorch(learning_rate)
+
+  # Models must implement a method to iterate all model.trainable_variables
```

Comment:
This is a hard constraint, it also makes this implementation very limited to yolo case. Is it possible to filter weights using regex patterns for weights, biases, and others? also do we need others? which learning rate is applied on others?

Here is an example for using regexp for weights: https://github.com/tensorflow/models/blob/master/official/modeling/optimization/configs/optimizer_config.py#L173

Reply:
yes others is needed. others is just another group, but weight decay is not applied to others. if the learning rate for a group is not set, by default it will use the same learning rate schedule as the weights. the bias group allow alternate schedules to be used for the biases.

Reply:
So this is essentially sgd with weight decay? In that case you're not building the model with l2 regularization layers, right?

Reply:
yes, l2 regularization is disabled in these cases.